### ISSUE TITLE
Support meta data in problem directories

### ISSUE THREAD

**Issue by jfilak** [abrt/libreport#354]
libreport was using two schemas for ownership and mode of dump directories in the past. The former was "abrt:user" + 0750 (used together with '/var/spool/abrt') and the latter is "user:abrt" + 0770 (used together with '/var/tmp/abrt'). However, neither of those schemas is good because the former allows any member of the user's group to read data from abrt problems of other users (and requires moving dump directories to user's home) and the latter must not be used together with system daemons because of symbolic  and hard link attacks and uncountable number of other security threats. The only possible schema acceptable by Security people is "root:abrt" + 0750 and the owner must never be changed ("abrt:abrt" is not acceptable because the owner must be root).

libreport uses the fs user/group to track the owner of the problem (not fs owner). With the change to "root:abrt", libreport cannot use fs attributes for this purpose any more. Thus I propose to add some kind of meta-data to dump directories.

Requirements:
* the meta-data must not be included in any kind of report (upload, Bugzilla, email)
* access to stored data must be really fast

Possible solutions:
1. files prefixed with '.' (".owner")
  * trivial to implement
  * extensible
  * error prone (the meta-data files would be mixed with the regular element files)
2. single file prefixed with '.' (".metadata" | ".libreport")
  * sounds like a good idea but requires fast parser (no JSON, no XML) and extensible format
3. single directory with some files (".metada" | ".libreport" / owner)
  * needs more coding
  * robust
  * extensible

I personally like the 3rd option. Do you have any comments or objections?

**Comment by mbrysa** [comment]
Which of 2. or 3. needs more coding in your opinion? Is most of the coding in 3. related to introducing the subdirectory?

**Comment by jfilak** [comment]
The libreport is based on dealing with plain text files, thus the 3rd option is much easier to implement as we already have functions for it. However, the 2nd option looks like a step in right direction, but I afraid of its complexity. Update a plain text file is incomparable easier then update a structured file. I wrote 'no JSON' but honestly if we decide for the 2nd, JSON is probably the best choice (even though I don't see any added value - the meta-data should be strictly internal).

**Comment by mbrysa** [comment]
I'm for the option 3.

**Comment by sorki** [comment]
Both 2 and 3 sounds good. IMO 2nd option would make sense if _all_ text elements are stored in json file but I don't see any benefits over current solution so 3rd option sounds good as well.

**Comment by jfilak** [comment]
I have patches for 3 ready, but I am tempted to drop them and go for 4 not-yet discussed option. Yesterday, *extended file attributes* came to my mind and I think *xattrs* is the right thing we need because *xattrs* is *metadata*. The security context is stored in the form of *extended file attributes*, so *xattrs* must be available in all target distributions. The implementation will ridiculously trivial -> `fsetxattr`/`fgetxattr` - (no parsing of text files, no sub directories, no pollution in the dump directory at all). What do you think?

**Comment by mbrysa** [comment]
Makes sense for me. We lose a bit of transparency, i.e. it might not be immediately obvious that that xattrs are used.

**Comment by jfilak** [comment]
Unfortunately. NFS does not support `xattrs` and there are users having abrt configured to store the problem data on NFS.
